recipe/linux-aarch64.patch

```diff
@@ -0,0 +1,12 @@
+--- tests/ndarray-eigen.cc.orig	2022-01-10 13:18:40.000000000 -0800
++++ tests/ndarray-eigen.cc	2022-01-10 13:19:20.000000000 -0800
+@@ -90,8 +90,8 @@
+     m1.setRandom(m, 6);
+     Eigen::Matrix<T, Eigen::Dynamic, Cols> m2(6, n);
+     m2.setRandom(6, n);
+-    asEigenMatrix(makeMutable(a)) = m1 * m2;
+     Eigen::Matrix<T, Rows, Cols> m3 = m1 * m2;
++    asEigenMatrix(makeMutable(a)) = m3;
```

Comment:
This change being necessary for aarch64 is very worrying; the original pattern (which avoids a memory allocation relative to the new one) is the recommended one, and I would expect it to be in heavy use.  Can you provide any details on what went wrong?

Reply:
Here's the output from these lines
```
diff --git a/tests/ndarray-eigen.cc b/tests/ndarray-eigen.cc
index 7d4785d..570c8f5 100644
--- a/tests/ndarray-eigen.cc
+++ b/tests/ndarray-eigen.cc
@@ -94,7 +94,7 @@ void testAsEigen2(A const & a) {
     Eigen::Matrix<T, Rows, Cols> m3 = m1 * m2;
     for (int i = 0; i < m; ++i) {
         for (int j = 0; j < n; ++j) {
-            BOOST_CHECK(a[i][j] == m3(i, j));
+            BOOST_TEST(a[i][j] == m3(i, j));
         }
     }
     Eigen::Array<T, Rows, Cols> m4(m, n);
```
```
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.62080642035384392 != 0.62080642035384404]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.098134428085439096 != 0.098134428085439082]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.54531818194560433 != 0.54531818194560444]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.72869435090084567 != 0.72869435090084556]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-0.37518403403472472 != -0.37518403403472467]
ndarray-eigen.cc(104): error: in "AsEigen": check (m5 == m6).all() has failed
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.26376205189687063 != 0.26376205189687052]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-0.61045022488560163 != -0.61045022488560152]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.17546361782477382 != 0.17546361782477379]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.34331900191979631 != 0.34331900191979625]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.68715163861550832 != 0.6871516386155081]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.62499860276599262 != 0.62499860276599273]
ndarray-eigen.cc(104): error: in "AsEigen": check (m5 == m6).all() has failed
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-0.37081844758208288 != -0.37081844758208282]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-0.73958926649480861 != -0.73958926649480849]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-0.041954928679806741 != -0.041954928679806734]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.10058109766045258 != 0.10058109766045253]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.0094112519709347309 != 0.0094112519709346181]
ndarray-eigen.cc(104): error: in "AsEigen": check (m5 == m6).all() has failed
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.11381474687443881 != 0.11381474687443885]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-0.80220189235959793 != -0.80220189235959805]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.10547973455071012 != 0.10547973455071014]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-0.1556093461439958 != -0.15560934614399577]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.31530568108775037 != 0.31530568108775048]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-0.9807395910583534 != -0.98073959105835362]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-1.368109569363035 != -1.3681095693630347]
ndarray-eigen.cc(104): error: in "AsEigen": check (m5 == m6).all() has failed
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-0.0723285675 != -0.072328575]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.0637856722 != 0.0637857765]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-0.042522639 != -0.0425226316]
ndarray-eigen.cc(104): error: in "AsEigen": check (m5 == m6).all() has failed
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-0.297502875 != -0.297502816]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [-0.855279922 != -0.855279863]
ndarray-eigen.cc(97): error: in "AsEigen": check a[i][j] == m3(i, j) has failed [0.397927999 != 0.39792797]
ndarray-eigen.cc(104): error: in "AsEigen": check (m5 == m6).all() has failed
```

Reply:
These lines are the the culprit for the output above:
   ```
    ndarray::Array<T, 2, 1> a21(a22[ndarray::view()(0, 3)]);
    testAsEigen2<5, 3, XprKind>(a21);
    testAsEigen2<3, 5, XprKind>(a21.transpose());

```

Reply:
The problem is in Eigen
` std::cout << ( m3 == m1 * m2) << std::endl;`
prints zero for the failing cases.

Reply:
Ok, so it's probably that one side of these comparisons is using some explicit vectorization while the other is not, due to compile-time knowledge about whether the memory is going to be aligned.  The patch here sort of voids the purpose of the test: to make sure that one can assign arbitrary Eigen lazy expression template objects to ndarray Eigen views, not just owning `Eigen::Matrix` instances.  It would be better to switch to an approximate comparison assertion, like `BOOST_CHECK_CLOSE`.